extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/search/filter/YoutubeProtoBufferSearchParameterAccessor.java

```diff
@@ -0,0 +1,227 @@
+
+import java.io.IOException;
+import java.net.URLDecoder;
+import java.net.URLEncoder;
+import java.util.ArrayList;
+import java.util.Base64;
+
+
+/**
+ * This class interacts with the auto generated proto buffer java files
+ * created by the 'squareup.wire proto buffer' plugin.
+ * <p>
+ * Below proto buffer description file is used:
+ * <a href="file:../main/proto/youtube-content-and-sort-filters.proto"
+ * >youtube-content-and-sort-filters.proto</a>
+ */
+public final class YoutubeProtoBufferSearchParameterAccessor {
+
+    private static final String UTF_8 = "UTF-8";
+
+    /**
+     * the base64 urlencoded sp string
+     */
+    private String searchParameter = "";
+
+    private YoutubeProtoBufferSearchParameterAccessor() {
+    }
+
+    @SuppressWarnings("NewApi")
+    public String encodeSp(final SortOrder sort, final DateFilter date,
+                           final TypeFilter type, final LengthFilter length,
+                           final Features[] features, final ExtraFeatures[] extraFeatures)
+            throws IOException {
+
+        final Filters.Builder filtersBuilder = new Filters.Builder();
+        if (null != date) {
+            filtersBuilder.date((long) date.getValue());
+        }
+
+        if (null != type) {
+            filtersBuilder.type((long) type.getValue());
+        }
+
+        if (null != length) {
+            filtersBuilder.length((long) length.getValue());
+        }
+
+        if (null != features) {
+            for (final Features feature : features) {
+                setFeatureState(feature, true, filtersBuilder);
+            }
+        }
+
+        final SearchRequest.Builder searchRequestBuilder = new SearchRequest.Builder();
+        if (null != sort) {
+            searchRequestBuilder.sorted((long) sort.getValue());
+        }
+
+        if (null != date || null != type || null != length || null != features
+                // even though extraFeatures is evaluated later. But in the case that extraFeatures
+                // will be the only activated 'feature' we still need to generate Filters here
+                // as it is integral part of the SearchRequest object
+                || null != extraFeatures) {
+            final Filters filters = filtersBuilder.build();
+            searchRequestBuilder.filter(filters);
+        }
+
+        if (null != extraFeatures && extraFeatures.length > 0) {
+            final Extras.Builder extrasBuilder = new Extras.Builder();
+            for (final ExtraFeatures extra : extraFeatures) {
+                setExtraState(extra, true, extrasBuilder);
+            }
+            final Extras extras = extrasBuilder.build();
+            searchRequestBuilder.extras(extras);
+        }
+
+        final SearchRequest searchRequest = searchRequestBuilder.build();
+
+        final byte[] protoBufEncoded = searchRequest.encode();
+        final String protoBufEncodedBase64 = Base64.getEncoder()
+                .encodeToString(protoBufEncoded);
```

Comment:
The `Base64` class was added to Android starting with API level 26 (Oreo), so this will crash on older devices.

Relevant issue: https://issuetracker.google.com/issues/171293709

Reply:
What's the alternative? Including Apache Common codec?

Reply:
I wonder if this PR is being used in the bravenewpipe fork which this PR originates from (which is a NewPipe fork)

Reply:
> What's the alternative? Including Apache Common codec?

There's a Base64 class included in Android.

Reply:
> > What's the alternative? Including Apache Common codec?
> 
> There's a Base64 class included in Android.
But NewPipeExtractor should not include android specific stuff. Am I right?

Reply:
> > > What's the alternative? Including Apache Common codec?
> > 
> > There's a Base64 class included in Android.
> But NewPipeExtractor should not include android specific stuff. Am I right?

Yeah, I updated my comment.